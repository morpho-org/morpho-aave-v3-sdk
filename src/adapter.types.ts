import {
  FetchedStatic,
  FetchedUpdated,
  MarketConfig,
  MarketData,
  MarketMapping,
  ScaledMarketData,
  ScaledMarketSupply,
  ScaledUserMarketData,
  UserMarketData,
} from "./types";

export type MarketsConfigs = MarketMapping<FetchedStatic<MarketConfig>>;
export type ScaledMarketsData = MarketMapping<
  FetchedStatic<ScaledMarketData & ScaledMarketSupply>
>;
export type MarketsData = MarketMapping<FetchedUpdated<MarketData>>;
export type UserMarketsData = MarketMapping<FetchedUpdated<UserMarketData>>;
export type ScaledUserMarketsData = MarketMapping<
  FetchedStatic<ScaledUserMarketData>
>;
