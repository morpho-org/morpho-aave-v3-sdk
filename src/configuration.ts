import { BigNumber } from "ethers";
import { parseUnits } from "ethers/lib/utils";

const defaultRpcHttpUrl = process.env.RPC_HTTP_URL || process.env.RPC_URL;

interface Configuration {
  readonly isProd: boolean;
  readonly rpcHttpUrl?: string;
  readonly network: number;
  readonly defaultMaxIterations: {
    readonly supply: number;
    readonly borrow: number;
  };
  readonly gasLimitPercent: BigNumber;
  readonly percentApproximation: BigNumber;
  readonly txSignature?: string;
}

class MorphoAaveV3Sdk {
  private __configuration: Configuration = {
    isProd: false,
    rpcHttpUrl: defaultRpcHttpUrl,
    network: 1,
    gasLimitPercent: parseUnits("1.1", 4),
    percentApproximation: parseUnits("0.99", 4), // 99%
    defaultMaxIterations: {
      supply: 4,
      borrow: 4,
    },
  } as const;

  public get configuration() {
    return { ...this.__configuration };
  }

  public setConfiguration = (newConfig: Partial<Configuration>) => {
    this.__configuration = { ...this.__configuration, ...newConfig };
  };
}

export default new MorphoAaveV3Sdk();
