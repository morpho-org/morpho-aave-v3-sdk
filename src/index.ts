import sdk from "./configuration";

export { MorphoAaveV3Adapter } from "./MorphoAaveV3Adapter";

export default sdk;
