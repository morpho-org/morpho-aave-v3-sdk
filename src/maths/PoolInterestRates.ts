import { BigNumber, constants } from "ethers";

import { SECONDS_PER_YEAR } from "../constants/date";

import { MorphoAaveMath } from "./AaveV3.maths";

export interface PoolIndexesParams {
  /** The current pool supply index (in ray). */
  lastPoolSupplyIndex: BigNumber;

  /** The current pool borrow index (in ray). */
  lastPoolBorrowIndex: BigNumber;

  /** The current pool supply rate (in ray). */
  poolSupplyRatePerYear: BigNumber;

  /** The current pool borrow rate (in ray). */
  poolBorrowRatePerYear: BigNumber;

  /** The last update timestamp (in seconds). */
  lastUpdateTimestamp: BigNumber;

  /** The current timestamp (in seconds). */
  currentTimestamp: BigNumber;
}

export default class PoolInterestRates {
  __MATHS__ = new MorphoAaveMath();

  /**
   * Recompute the exact same logic as the Aave V3 protocol.
   * For supply index: https://github.com/aave/aave-v3-core/blob/9630ab77a8ec77b39432ce0a4ff4816384fd4cbf/contracts/protocol/libraries/logic/ReserveLogic.sol#L47
   * For borrow index: https://github.com/aave/aave-v3-core/blob/9630ab77a8ec77b39432ce0a4ff4816384fd4cbf/contracts/protocol/libraries/logic/ReserveLogic.sol#L73
   */
  public computePoolIndexes({
    lastPoolSupplyIndex,
    lastPoolBorrowIndex,
    poolSupplyRatePerYear,
    poolBorrowRatePerYear,
    lastUpdateTimestamp,
    currentTimestamp,
  }: PoolIndexesParams) {
    if (lastUpdateTimestamp.gte(currentTimestamp))
      return {
        newPoolSupplyIndex: lastPoolSupplyIndex,
        newPoolBorrowIndex: lastPoolBorrowIndex,
      };

    const newPoolSupplyIndex = this.__MATHS__.indexMul(
      lastPoolSupplyIndex,
      this._calculateLinearInterest(
        poolSupplyRatePerYear,
        lastUpdateTimestamp,
        currentTimestamp
      )
    );

    const newPoolBorrowIndex = this.__MATHS__.indexMul(
      lastPoolBorrowIndex,
      this._calculateCompoundedInterest(
        poolBorrowRatePerYear,
        lastUpdateTimestamp,
        currentTimestamp
      )
    );

    return {
      newPoolSupplyIndex,
      newPoolBorrowIndex,
    };
  }

  private _calculateLinearInterest(
    rate: BigNumber,
    lastUpdateTimestamp: BigNumber,
    currentTimestamp: BigNumber
  ) {
    const exp = currentTimestamp.sub(lastUpdateTimestamp);

    if (exp.isZero()) return this.__MATHS__.INDEX_ONE;

    return this.__MATHS__.INDEX_ONE.add(rate.mul(exp).div(SECONDS_PER_YEAR));
  }
  private _calculateCompoundedInterest(
    rate: BigNumber,
    lastUpdateTimestamp: BigNumber,
    currentTimestamp: BigNumber
  ) {
    const exp = currentTimestamp.sub(lastUpdateTimestamp);

    if (exp.isZero()) return this.__MATHS__.INDEX_ONE;

    const expMinusOne = exp.sub(1);
    const expMinusTwo = exp.gt(2) ? exp.sub(2) : constants.Zero;

    const basePowerTwo = this.__MATHS__
      .indexMul(rate, rate)
      .div(BigNumber.from(SECONDS_PER_YEAR).mul(SECONDS_PER_YEAR));

    const basePowerThree = this.__MATHS__
      .indexMul(basePowerTwo, rate)
      .div(SECONDS_PER_YEAR);

    const secondTerm = exp.mul(expMinusOne).mul(basePowerTwo).div(2);

    const thirdTerm = exp
      .mul(expMinusOne)
      .mul(expMinusTwo)
      .mul(basePowerThree)
      .div(6);

    return this.__MATHS__.INDEX_ONE.add(rate.mul(exp).div(SECONDS_PER_YEAR))
      .add(secondTerm)
      .add(thirdTerm);
  }
}
