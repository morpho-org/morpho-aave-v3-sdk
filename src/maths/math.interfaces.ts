import { BigNumberish, BigNumber } from "ethers";

export interface IMath {
  indexMul: (value: BigNumberish, index: BigNumberish) => BigNumber;
  indexDiv: (value: BigNumberish, index: BigNumberish) => BigNumber;
  INDEX_ONE: BigNumber;

  mul: (a: BigNumber, b: BigNumber) => BigNumber;
  div: (a: BigNumber, b: BigNumber) => BigNumber;
  ONE: BigNumber;

  percentMul: (value: BigNumber, pct: BigNumber) => BigNumber;
  percentDiv: (value: BigNumber, pct: BigNumber) => BigNumber;
  PERCENT_ONE: BigNumber;

  computeApysFromRates: (
    poolSupplyRate: BigNumber,
    poolBorrowRate: BigNumber,
    p2pIndexCursor: BigNumber,
    reserveFactor: BigNumber,
    proportionDelta: BigNumber,
    proportionIdle: BigNumber
  ) => {
    poolBorrowAPY: BigNumber;
    poolSupplyAPY: BigNumber;
    p2pSupplyAPY: BigNumber;
    p2pBorrowAPY: BigNumber;
  };
}
