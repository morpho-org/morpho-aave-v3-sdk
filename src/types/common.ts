import { BigNumber } from "ethers";

import { Block, FeeData } from "@ethersproject/abstract-provider";

export type Address = string;

export interface GlobalData {
  /** The current block. All chain data should be up to date with this block */
  readonly currentBlock: Block;

  /** Timestamp of the last fetch */
  readonly lastFetchTimestamp: number;

  /** ETH Price in usd
   *
   * Number of decimals:
   * `18` _(WEI)_
   */
  readonly ethUsdPrice: BigNumber;

  /** Data related to fees */
  readonly feeData: FeeData;

  /** Configuration of the selected eMode */
  readonly eModeCategoryData: EModeCategoryData;

  /** Current rewards distributor root */
  readonly currRoot: string;
}
export interface EModeCategoryData {
  /** Unique id of the emode */
  eModeId: BigNumber;

  /** Common Loan To Value for all markets in this eMode
   *
   * _NB: This overrides the initial market's LTV_
   *
   * Number of decimals:
   * `18` _(WEI)_
   */
  ltv: BigNumber;

  /** Common Liquidation Threshold for all markets in this eMode
   *
   * _NB: This overrides the initial market's LT_
   *
   * Number of decimals:
   * `18` _(WEI)_
   */
  liquidationThreshold: BigNumber;

  /** Common Liquidation Bonus for all markets in this eMode
   *
   * _NB: This overrides the initial market's liquidation bonus_
   *
   * Number of decimals:
   * `18` _(WEI)_
   */
  liquidationBonus: BigNumber;

  /** Contract used to fetch the price of all the markets in the eMode
   *
   * _NB: The price fetched using this source is overriding the initial price of the market_
   *
   * Number of decimals:
   * `18` _(WEI)_
   */
  priceSource: Address;

  /** Label of the eMode */
  label: string;
}

export enum MaxCapacityLimiter {
  walletBalance = "LIMITED_BY_WALLET_BALANCE",
  operationPaused = "LIMITED_BY_OPERATION_PAUSED",
  zeroPrice = "LIMITED_BY_ZERO_PRICE",
  borrowCapacity = "LIMITED_BY_BORROW_CAPACITY",
  poolLiquidity = "LIMITED_BY_POOL_LIQUIDITY",
  cap = "LIMITED_BY_CAP",
  balance = "LIMITED_BY_BALANCE",
}

export interface MaxCapacity {
  /** Maximum inputable amount for the given operation (in underlying) */
  amount: BigNumber;
  /** Reason limiting the maximum inputable amount */
  limiter: MaxCapacityLimiter;
}
