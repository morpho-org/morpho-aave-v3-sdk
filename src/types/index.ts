export * from "./generics";
export * from "./common";
export * from "./markets";
export * from "./user";
export * from "./tx";
