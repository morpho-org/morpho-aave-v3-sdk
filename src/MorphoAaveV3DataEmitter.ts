import { deepCopy } from "ethers/lib/utils";
import { Subject } from "rxjs";

import { MorphoAaveV3DataHolder } from "./MorphoAaveV3DataHolder";
import { MarketsConfigs, MarketsData, UserMarketsData } from "./adapter.types";
import { FetchedStatic, FetchedUpdated, GlobalData, UserData } from "./types";

export abstract class MorphoAaveV3DataEmitter extends MorphoAaveV3DataHolder {
  /* Subjects */
  public readonly marketsConfigs$ = new Subject<MarketsConfigs>();
  public readonly marketsData$ = new Subject<MarketsData>();
  public readonly userMarketsData$ = new Subject<UserMarketsData>();
  public readonly marketsList$ = new Subject<FetchedStatic<string[]>>();
  public readonly userData$ = new Subject<FetchedUpdated<UserData>>();
  public readonly globalData$ = new Subject<FetchedUpdated<GlobalData>>();

  /* Setters */
  protected set marketsConfigs(mc: MarketsConfigs) {
    this._marketsConfigs = mc;
    this.marketsConfigs$.next(deepCopy(mc));
  }
  protected set marketsData(md: MarketsData) {
    this._marketsData = md;
    this.marketsData$.next(deepCopy(md));
  }
  protected set userMarketsData(umd: UserMarketsData) {
    this._userMarketsData = umd;
    this.userMarketsData$.next(deepCopy(umd));
  }
  protected set marketsList(ml: FetchedStatic<string[]>) {
    this._marketsList = ml;
    this.marketsList$.next(deepCopy(ml));
  }
  protected set userData(ud: FetchedUpdated<UserData>) {
    this._userData = ud;
    this.userData$.next(deepCopy(ud));
  }
  protected set globalData(gd: FetchedUpdated<GlobalData>) {
    this._globalData = gd;
    this.globalData$.next(deepCopy(gd));
  }
}
