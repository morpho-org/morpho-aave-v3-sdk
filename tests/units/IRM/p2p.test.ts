import { BigNumber, constants } from "ethers";

import { WadRayMath } from "@morpho-labs/ethers-utils/lib/maths";

import P2PInterestRates, {
  IndexesParams,
  MarketSizeIndexes,
} from "../../../src/maths/P2PInterestRates";
import { Deltas } from "../../../src/types";

describe("P2PInterestRates", () => {
  let p2pInterestRates: P2PInterestRates;

  beforeEach(() => {
    p2pInterestRates = new P2PInterestRates();
  });

  describe("computeP2PIndexes", () => {
    let indexesParams: IndexesParams;
    let lastSupplyIndexes: MarketSizeIndexes;
    let lastBorrowIndexes: MarketSizeIndexes;
    beforeEach(() => {
      // Default params values used as a reference in multiple tests
      lastSupplyIndexes = {
        poolIndex: WadRayMath.parseRay("1"),
        p2pIndex: WadRayMath.parseRay("1"),
      };
      lastBorrowIndexes = {
        poolIndex: WadRayMath.parseRay("1"),
        p2pIndex: WadRayMath.parseRay("1"),
      };
      indexesParams = {
        lastSupplyIndexes,
        lastBorrowIndexes,
        poolSupplyIndex: WadRayMath.parseRay("1.5"),
        poolBorrowIndex: WadRayMath.parseRay("2"), // borrow growth rate is higher than supply one
        reserveFactor: constants.Zero,
        p2pIndexCursor: BigNumber.from("5000"), // 50%
        deltas: {
          supply: {
            scaledDelta: constants.Zero,
            scaledP2PTotal: constants.Zero,
          },
          borrow: {
            scaledDelta: constants.Zero,
            scaledP2PTotal: constants.Zero,
          },
        },
        proportionIdle: constants.Zero,
      };
    });

    describe("in case deltas are equals to 0", () => {
      let deltas: Deltas;
      beforeAll(() => {
        // 0 deltas
        deltas = {
          supply: {
            scaledDelta: constants.Zero,
            scaledP2PTotal: constants.Zero,
          },
          borrow: {
            scaledDelta: constants.Zero,
            scaledP2PTotal: constants.Zero,
          },
        };
      });

      /*
       * We play with different values of p2pIndexCursor and reserveFactor
       * which should not change the expected result if both the supply and the borrow index
       * are not increased
       */
      it("should return same indexes when pool indexes didn't change", () => {
        indexesParams = {
          ...indexesParams,
          poolSupplyIndex: lastSupplyIndexes.poolIndex,
          poolBorrowIndex: lastBorrowIndexes.poolIndex,
        };

        // crossing different values
        ["10000", "7000", "5000", "0"].forEach((p2pIndexCursor) => {
          ["10000", "7345", "5000", "0"].forEach((reserveFactor) => {
            const { newP2PSupplyIndex, newP2PBorrowIndex } =
              p2pInterestRates.computeP2PIndexes({
                ...indexesParams,
                p2pIndexCursor: BigNumber.from(p2pIndexCursor),
                reserveFactor: BigNumber.from(reserveFactor),
              });

            expect(newP2PSupplyIndex).toBnEq(lastSupplyIndexes.p2pIndex);
            expect(newP2PBorrowIndex).toBnEq(lastBorrowIndexes.p2pIndex);
          });
        });
      });

      describe("REPL test cases", () => {
        /*
         * We test a 'normal' case
         */
        it("CASE 1: p2pIndex at 50% and supply growth rate <= borrow growth rate ", () => {
          // Saving a snapshot of indexes Params at first execution.
          expect(indexesParams).toMatchSnapshot();
          // We then have to copy those values to the test_cases.sol file (CASE 1)
          // and run the case code it in REPL to compute the expected values.

          // REPL CASE 1 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1750000000000000000000000000"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1750000000000000000000000000"
            ),
          };

          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });

        /*
         * We simulate a case where the supply grows faster than the borrow
         * https://github.com/morpho-dao/morpho-aave-v3/blob/main/src/libraries/InterestRatesLib.sol#L79
         */
        it("CASE 2: supply index should be limited by borrow index growth", () => {
          indexesParams = {
            ...indexesParams,
            poolSupplyIndex: WadRayMath.parseRay("1.1"),
            poolBorrowIndex: WadRayMath.parseRay("1.05"), // borrow growth rate is lower than supply
            p2pIndexCursor: constants.Zero,
          };
          expect(indexesParams).toMatchSnapshot();

          // REPL CASE 2 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1050000000000000000000000000"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1050000000000000000000000000"
            ),
          };
          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });

        /*
         * This should have no impact has proportionIdle only comes into play if scaledP2PTotal != 0
         * https://github.com/morpho-dao/morpho-aave-v3/blob/8c5bdf9e349f3cd9da122adde1eedc6781542352/src/libraries/InterestRatesLib.sol#L107
         */
        it("CASE 3: reserve factor and p2pIndexCursor non zero", () => {
          const lastSupplyIndexes = {
            poolIndex: WadRayMath.parseRay("1"),
            p2pIndex: WadRayMath.parseRay("1"),
          };
          const lastBorrowIndexes = {
            poolIndex: WadRayMath.parseRay("1"),
            p2pIndex: WadRayMath.parseRay("1"),
          };

          const indexesParams = {
            lastSupplyIndexes,
            lastBorrowIndexes,
            poolSupplyIndex: WadRayMath.parseRay("1.1"),
            poolBorrowIndex: WadRayMath.parseRay("1.3"),
            reserveFactor: BigNumber.from("1000"), // 10%
            p2pIndexCursor: BigNumber.from("2237"), // 22.37%
            deltas,
            proportionIdle: constants.Zero,
          };
          expect(indexesParams).toMatchSnapshot();

          // REPL CASE 3 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1140266000000000000000000000"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1160266000000000000000000000"
            ),
          };
          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });
        it("CASE 4: proportion idle non zero", () => {
          indexesParams = {
            ...indexesParams,
            proportionIdle: BigNumber.from("1000"), // 10%
          };
          // Saving a snapshot of indexes Params at first execution.
          expect(indexesParams).toMatchSnapshot();

          // REPL CASE 4 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1750000000000000000000000000"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1750000000000000000000000000"
            ),
          };
          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });
      }); // REPL test cases
    }); // deltas equals to 0

    describe("in case deltas non zero", () => {
      let deltas: Deltas;
      beforeAll(() => {
        // 0 deltas
        deltas = {
          supply: {
            scaledDelta: WadRayMath.parseWad("1000"),
            scaledP2PTotal: WadRayMath.parseWad("9000"),
          },
          borrow: {
            scaledDelta: WadRayMath.parseWad("10"),
            scaledP2PTotal: WadRayMath.parseWad("9010"),
          },
        };
      });
      describe("REPL test cases", () => {
        /*
         * non zeros delta induce the calculation of the proportionDelta in InterestRatesLib.sol:computeP2PIndex()
         * https://github.com/morpho-dao/morpho-aave-v3/blob/8c5bdf9e349f3cd9da122adde1eedc6781542352/src/libraries/InterestRatesLib.sol#L105
         */
        it("CASE A1: proportionDelta computation in computeP2PIndex()", () => {
          indexesParams = {
            ...indexesParams,
            deltas,
          };
          expect(indexesParams).toMatchSnapshot();

          // REPL CASE A1 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1722222222222222222222222222"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1750277469478357380688124307"
            ),
          };
          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });
        it("CASE A2 proportion idle", () => {
          indexesParams = {
            ...indexesParams,
            deltas,
            proportionIdle: BigNumber.from("1000"),
          };
          expect(indexesParams).toMatchSnapshot();

          // REPL CASE A2 solidity code execution result (check IRM/README.md for more info):
          const { expectedNewP2PSupplyIndex, expectedNewP2PBorrowIndex } = {
            expectedNewP2PSupplyIndex: BigNumber.from(
              "1722222222222222222222221472"
            ),
            expectedNewP2PBorrowIndex: BigNumber.from(
              "1750277469478357380688124307"
            ),
          };

          const { newP2PSupplyIndex, newP2PBorrowIndex } =
            p2pInterestRates.computeP2PIndexes(indexesParams);

          expect(newP2PSupplyIndex).toBnEq(expectedNewP2PSupplyIndex);
          expect(newP2PBorrowIndex).toBnEq(expectedNewP2PBorrowIndex);
        });
      });
    }); // deltas npn zeros
  }); // computeP2PIndexes
});
