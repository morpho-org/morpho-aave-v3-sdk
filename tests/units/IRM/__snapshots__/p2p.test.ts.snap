// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`P2PInterestRates computeP2PIndexes in case deltas are equals to 0 REPL test cases CASE 1: p2pIndex at 50% and supply growth rate <= borrow growth rate  1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x1388",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x06765c793fa10079d0000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x04d8c55aefb8c05b5c000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x00",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
}
`;

exports[`P2PInterestRates computeP2PIndexes in case deltas are equals to 0 REPL test cases CASE 2: supply index should be limited by borrow index growth 1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x03648a260e3486a65a000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x038de60f7c988d0fcc000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x00",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
}
`;

exports[`P2PInterestRates computeP2PIndexes in case deltas are equals to 0 REPL test cases CASE 3: reserve factor and p2pIndexCursor non zero 1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x08bd",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x043355b53628a6b594000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x038de60f7c988d0fcc000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x00",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x03e8",
    "type": "BigNumber",
  },
}
`;

exports[`P2PInterestRates computeP2PIndexes in case deltas are equals to 0 REPL test cases CASE 4: proportion idle non zero 1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x00",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x1388",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x06765c793fa10079d0000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x04d8c55aefb8c05b5c000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x03e8",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
}
`;

exports[`P2PInterestRates computeP2PIndexes in case deltas non zero REPL test cases CASE A1: proportionDelta computation in computeP2PIndex() 1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x8ac7230489e80000",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x01e86ede3ef95d880000",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x3635c9adc5dea00000",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x01e7e4171bf4d3a00000",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x1388",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x06765c793fa10079d0000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x04d8c55aefb8c05b5c000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x00",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
}
`;

exports[`P2PInterestRates computeP2PIndexes in case deltas non zero REPL test cases CASE A2 proportion idle 1`] = `
{
  "deltas": {
    "borrow": {
      "scaledDelta": {
        "hex": "0x8ac7230489e80000",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x01e86ede3ef95d880000",
        "type": "BigNumber",
      },
    },
    "supply": {
      "scaledDelta": {
        "hex": "0x3635c9adc5dea00000",
        "type": "BigNumber",
      },
      "scaledP2PTotal": {
        "hex": "0x01e7e4171bf4d3a00000",
        "type": "BigNumber",
      },
    },
  },
  "lastBorrowIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "lastSupplyIndexes": {
    "p2pIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
    "poolIndex": {
      "hex": "0x033b2e3c9fd0803ce8000000",
      "type": "BigNumber",
    },
  },
  "p2pIndexCursor": {
    "hex": "0x1388",
    "type": "BigNumber",
  },
  "poolBorrowIndex": {
    "hex": "0x06765c793fa10079d0000000",
    "type": "BigNumber",
  },
  "poolSupplyIndex": {
    "hex": "0x04d8c55aefb8c05b5c000000",
    "type": "BigNumber",
  },
  "proportionIdle": {
    "hex": "0x03e8",
    "type": "BigNumber",
  },
  "reserveFactor": {
    "hex": "0x00",
    "type": "BigNumber",
  },
}
`;
